File: day2/ruby/puzzle2.rb
data = File.read File.join(__dir__, '/../input.txt')


def compare_distance a, b
  (0..a.length).inject(0) do |distance, idx|
    distance += 1 if a[idx] != b[idx]
    distance
  end
end

def common_letters a, b
  (0..a.length).inject([]) do |letters, idx|
    letters.push(a[idx]) if a[idx] == b[idx]
    letters
  end
end

boxes = data.split("\n")
box_count = boxes.length

(0..box_count-2).each do |idx1|
  (idx1+1..box_count-1).each do |idx2|
    if compare_distance(boxes[idx1], boxes[idx2]) == 1
      in_common = common_letters boxes[idx1], boxes[idx2]

      puts "Boxes with distance of 1: #{boxes[idx1]} and #{boxes[idx2]}"
      puts "Common letters: #{in_common.join('')}"
      exit
    end
  end
end

File: day2/python/puzzle2.py
import os

def compare_distance(a, b):
    distance = 0
    for idx in range(len(a)):
        if a[idx] != b[idx]:
            distance += 1

    return distance

def common_letters(a, b):
    return [a[idx] for idx in range(len(a)) if a[idx] == b[idx]]

dir_path = os.path.dirname(os.path.realpath(__file__))
data = []

with open(dir_path + "/../input.txt", "r") as datafile:
    for line in datafile:
        data.append(line.rstrip("\n"))

for idx1 in range(0, len(data)-2):
    for idx2 in range(idx1+1, len(data)-1):
        if compare_distance(data[idx1], data[idx2]) == 1:
            in_common = "".join(common_letters(data[idx1], data[idx2]))
            print(f"Boxes with distance of 1: {data[idx1]} and {data[idx2]}")
            print(f"Common letters: {in_common}")
